lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,250 @@
+          url_join('file:/', path)
+        end
+
+        # Retrive the content of a local or remote file.
+        #
+        # @param A local or remote uri.
+        #
+        # @raise ArgumentError if it's not a valid uri
+        #
+        # Protect open-uri from malign arguments like
+        #  - "|ls"
+        #  - multiline strings
+        def get_content(uri_or_filename)
+          remote_schemes = %w(http https ftp)
+          Fog::Logger.debug("Opening #{uri_or_filename}")
+
+          begin
+            # Validate URI to protect from open-uri attacks.
+            url = URI(uri_or_filename)
+
+            # Remote schemes must contain an host.
+            raise ArgumentError if url.host.nil? && remote_schemes.include?(url.scheme)
+
+            # Encode URI with spaces.
+            uri_or_filename = URI.encode(URI.decode(URI(uri_or_filename).to_s))
+          rescue URI::InvalidURIError
+            raise ArgumentError, "Not a valid URI: #{uri_or_filename}"
+          end
+
+          # TODO: A future revision may implement a retry.
+          # TODO: A future revision may limit download size.
+          content = ''
+          # open-uri doesn't open "file:///" uris.
+          uri_or_filename = uri_or_filename.sub(%r/file:/, "")
+
+          open(uri_or_filename) { |f| content = f.read }
+          content
+        end
+
+        # Retrieve a template content.
+        #
+        # @param template_file can be either:
+        #          - a raw_template string
+        #          - an URI string
+        #          - an Hash containing the parsed template.
+        #
+        # XXX: after deprecation of Ruby 1.9 we could use
+        #      named parameters and better mimic heatclient implementation.
+        def get_template_contents(template_file)
+          Fog::Logger.warning("get_template_contents #{template_file}")
+
+          raise "template_file should be Hash or String" unless
+            template_file.kind_of?(String) || template_file.kind_of?(Hash)
+
+          local_base_url = base_url_for_url(normalise_file_path_to_url(Dir.pwd + "/TEMPLATE"))
+
+          if template_file.kind_of?(Hash)
+            template_base_url = local_base_url
+            raw_template = YAML.dump(template_file)
+          elsif template_is_raw?(template_file)
+            raw_template = template_file
+            template_base_url = local_base_url
+          elsif template_is_url?(template_file)
+            template_file = normalise_file_path_to_url(template_file)
+            template_base_url = base_url_for_url(template_file)
+            raw_template = get_content(template_file)
+
+            Fog::Logger.warning("Template visited: #{@visited}")
+            @visited[template_file] = true
+          else
+            raise "template_file is not a string of the expected form"
+          end
+          template = yaml_load(raw_template)
+
+          get_file_contents(template, template_base_url)
+
+          return nil, template
```

Comment:
Why return `nil, template`. You can return only `template`, which is expected from the method name too.

First of all, I don't understand in what scenario can the input be a Hash.
Secondly, if the input is already a Hash, you don't need to convert it to `raw_template` and then convert it back to Hash.

Reply:
@bzwei 

- removed `nil`
- you can either pass a template as an Hash or a String
- I round-trip the Hash to deep-copy it in another object and don't modifying the passed argument. See https://github.com/ioggstream/fog-openstack/blob/218ee9b905b8a2e265c1e3bdba65e55da01db647/test/requests/orchestration/stack_files_util_tests.rb#L106